### ISSUE TITLE
Order-dependence setting nested arrays of different shapes

### ISSUE THREAD

**Issue by comatrion** [numpy/numpy#9684]
For example:
```
>>> a = np.array([np.array([1, 2]), np.array(1)])
>>> a
array([array([1, 2]), array(1)], dtype=object)

>>> b = np.array([np.array(1), np.array([1, 2])])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: setting an array element with a sequence.
```

Specifying dtype avoids the exception for b:
```
>>> b = np.array([np.array(1), np.array([1, 2])], dtype=object)
>>> b
array([array(1), array([1, 2])], dtype=object)
```
However dtype is optional. If an array can be instantiated successfully without dtype for a given input sequence, then re-ordering the sequence should not cause an exception.

python 3.4.5 numpy 1.13.1.

**Comment by jsclose** [comment]
Is this an actual issue? Should numpy automatically re order the array?

**Comment by eric-wieser** [comment]
We're definitely aware of this, and there's an issue somewhere suggesting that we change this behavior so that passing `dtype=object` is always required, and that otherwise both cases should error.

**Comment by jsclose** [comment]
[#5353](https://github.com/numpy/numpy/issues/5353) it seems like issue and this experiment [#10147](https://github.com/numpy/numpy/pull/10147). Was there any resolution to that?

**Comment by eric-wieser** [comment]
Fixed in #11601
